Out-of-band data + SSH Agent forwarding (Take 2)

kevinr opened the pull request:
I've cleaned up the compiler warnings in Timo Rinne's [ssh-agent-forwarding branch](https://github.com/rinne/mosh/tree/ssh-agent-forwarding branch) from pull request #423, to address issue #120, and squashed the change down into a single commit.  (It looks like you weren't configuring with `--enable-compile-warnings=error`, Timo, which the README recommends for distributors, and otherwise the Make output suppresses warnings.)

In my limited testing on Ubuntu Trusty, the patch works as advertised.

I'd love to see this merged to master.

luto commented on src/agent/agent.h:
```diff
@@ -0,0 +1,123 @@
+    void notify_eof(uint64_t agent_id);
+    AgentConnection *get_session();
+
+    ProxyAgent(const ProxyAgent&);    // unimplemented
```
are those two still unimplemented?

kevinr replied:
Yes -- the copy constructor and operator= method aren't used, but the class has pointer members so leaving them undefined produces a warning.  Defining them as private but not specifying them results in a link error if someone tries to use them anywhere.

luto replied:
mkay :) running out of C++ knowledge here :D

luto commented on src/frontend/stmclient.cc:
```diff
@@ -126,7 +127,7 @@ void STMClient::init( void )
     if ( strlen( escape_key_env ) == 1 ) {
       escape_key = (int)escape_key_env[0];
-      if ( escape_key > 0 && escape_key < 128 ) {
+      if ( escape_key > 0 || escape_key < 128 ) {
```
`||`? so basically any char value is allowed?

kevinr replied:
Good catch, @luto. What purpose this serves is a question for @rinne, the patch's original author. Blame says it comes from a49aef81.  Might be a mis-merge?

rinne replied:
It should be &&.

Meaning that any single character ASCII value is accepted (except NUL) and if the environment contains something else (e.g. a ISO-8859-1 character, which also could be a start sequence of UTF-8 combined character) it's just silently ignored. Then the magic is to separate ASCII-characters below 32 so that they can be referred as CTRL-something in info strings rather than by the character value itself.

BTW, there is still one small bug in escape character info string.  If the connection is offline (e.g. network connection goes broken) and mosh starts to show the info string about how to kill the process. That doesn't show the correct escape string if it has been modified by the environment but it rather says always CTRL-^. I've been thinking about fixing it sometimes but always forgotten. It is probably a one or two-liner, so please feel free.

rinne replied:
And in addition to one character ascii value, empty environment (explicitly set to empty string) disables the escape character altogether. Environment of any other kind, is just ignored.

kevinr replied:
Fixed to &&. Thanks for the information @rinne!

luto commented:
Thanks for the work! :+1:
Even though I'm neither a C++ programmer nor a mosh maintainer, I've added some comments to help @keithw out a bit :P

kevinr commented:
Thanks @luto !

I'm sorry @rinne, I didn't manage to tag you when I created this pull request.  I cleaned up the compiler warnings from your branch and squashed the commits down.

rinne commented:
@kevinr No problem. I'm just glad someone is willing to catch that, since I really don't have time for it and it has passed the magical "works for me" -level.

Kriechi commented:
:+1: 
works for me on OS X 10.9 client with Debian 7.8 as server.

Dorthu commented:
:+1: Working on OSX/Debian 7

moltar commented:
yay.

toresbe commented:
Awesome! Thanks. This makes my day so much easier when maintaining servers on the other side of the planet. Confirmed working on Debian 8 going to FreeBSD.

jravetch commented:
:+1:

nathanlburns commented:
Is there any update here? Additional testing needed?

Mic92 commented:
Thanks for this feature, for those on archlinux I have upload the package mosh-sshagent-git in the AUR based on the kevinr:kevinr-ssh-agent-forwarding branch.

PAStheLoD commented:
Hello,

You might be interested in knowing, that there is a problem with compiling this with g++ 4.9.

```
g++ -DHAVE_CONFIG_H -I. -I../..  -I./../util -I./../crypto -I../protobufs -pthread   -Wall -Werror -Wextra -pedantic -Wno-long-long -Weffc++ -fno-strict-overflow -D_FORTIFY_SOURCE=2 -fstack-protector-all -Wstack-protector --param ssp-buffer-size=1 -fPIE -fno-default-inline -pipe -g -O2 -c -o networktransport.o networktransport.cc
In file included from transportsender.h:45:0,
                 from networktransport.h:43,
                 from networktransport.cc:35:
outofband.h:118:9: error: 'class Network::OutOfBandPlugin' has virtual functions and accessible non-virtual destructor [-Werror=non-virtual-dtor]
   class OutOfBandPlugin
         ^
cc1plus: all warnings being treated as errors
```
 
Interestingly g++ 4.8 works fine.

glensc commented:
the warning can be non-fatal: `-Wno-error=non-virtual-dtor`

MacGyverNL commented:
I've started using gpg-agent-forwarding through the SSH-mechanism for arbitrary Unix Domain socket forwarding (ssh -R /home/remoteuser/.gnupg/S.gpg-agent /home/localuser/.gnupg/S.forwarded-agent, recent addition to OpenSSH). I have not looked at the code in OpenSSH used to do this, but looking at the patches for SSH-agent-forwarding it seems to me that support for forwarding arbitrary Unix Domain sockets should not be *too* different from what's here now. Would this be an interesting thing to add?

rinne commented:
The current implementation of agent forwarding actually parses the content of the stream so that it parses and sanity checks protocol packet length and separates the packets from each other (does not care about packet contents). This is done in order to keep the flow control issues in minimum. Adding specific gpg agent forwarding would probably be quite easy and it could borrow most of the ssh agent forwarding code and run over the out-of-band transport like it. But adding a generic forwarding with full socket flow control would be _much_ bigger effort.

glensc commented:
@kevinr would you mind rebasing it to current master? so could build  this patch against [1.2.4.95rc1 pre-release](#625)

cgull commented on src/agent/agent.cc:
```diff
@@ -0,0 +1,490 @@
+    return;
+  }
+  if (server) {
+    PRNG prng;
```
This really is close to being a bikeshed...but I have to agree that `mkdtemp()` is a better thing to do here.  Your code *looks* OK, but the people who write libc code have thought longer and harder on this topic than you or I have.

rinne replied:
This is pretty much a non-issue, but I have a recollection that there really is a valid reason not to use library function here. Also this is really simple stuff.

cgull commented on src/frontend/mosh-server.cc:
```diff
@@ -551,6 +576,8 @@ void serve( int host_fd, Terminal::Complete &terminal, ServerConnection &network
 	sel.add_fd( host_fd );
       }
 
+      network.oob()->pre_poll();
```
Having all these calls to network.oob, and the pre/post machinery, feels awkward to me.  It seems to me that this is all a workaround for having `network`'s `select` calls here at top level.  Perhaps we can push the selects down into `Network::Transport`, and have that select for and call out to the OOB object (or maybe even push the select object down another level to OOB as well).

Adding another set of data flows to Mosh is going to be a little awkward no matter how it's done, but I think it can be a little less awkward than this :)

rinne replied:
I think going this other way around, would be quite a bit of (re-)work. But as long as I don't have to do it, I don't mind.

cgull commented on src/util/swrite.cc:
```diff
@@ -53,3 +59,42 @@ int swrite( int fd, const char *str, ssize_t len )
 }
+
+
+int swrite_timeout( int fd, uint64_t timeout_ms, const char *str, ssize_t len )
```
I don't think this (and the code that calls it) is workable.  This is a blocking call inside Mosh.

Mosh is (imperfectly trying to be) an event driven program; it should accept any input and act on it, or generate any output when its file descriptor is ready.  This function blocks the rest of Mosh while waiting to complete the write.

There's an elephant already in the room here, though: Mosh can also block on the tty it's attached to (and block it by not responding fast enough, as well).  I just want to avoid adding another one.

rinne replied:
Well, the sole reason for it being there, is the blocking socket write (without timeout) that already was just next to it. While not very elegant, the reasoning there was that should agent connection in the client end block, it should after a reasonably short timeout just fail so that the process that tries to use it via forwarding, can continue its authentication process using other means.
If someone wishes to write this in some other way, please do, but that's more or less workable.

cgull commented on src/agent/agent.cc:
```diff
@@ -0,0 +1,490 @@
+  return (Network::timestamp() - idle_start) / 1000;
+}
+
+string AgentConnection::recv_packet() {
```
This method has an issue: it assumes the entire agent-forwarding message is immediately available.  If it were to encounter a client that does unbuffered i/o and does separate writes for the message length and message data, it would fail unpredictably.  This should be reworked to an event driven style.

rinne replied:
What? No it doesn't. At least it's not supposed to do it. What it does though, it parses the packet length once it has it and kills entire agent session (meaning a single connection from the user program to the agent server) if the packet seems to be thrash (i.e. too big or zero length). However through mosh oob protocol, the agent packets are sent as entire packets.

cgull commented on src/frontend/mosh-server.cc:
```diff
@@ -185,8 +189,11 @@ int main( int argc, char *argv[] )
     /* new option syntax */
     int opt;
-    while ( (opt = getopt( argc - 1, argv + 1, "i:p:c:svl:" )) != -1 ) {
+    while ( (opt = getopt( argc - 1, argv + 1, "i:p:c:svl:A" )) != -1 ) {
```
It'd be nice if we didn't need another option here, and could learn this by sending the option in the packets.  But the server starts the child shell immediately, and needs to know whether or not to set SSH_AUTH_SOCK.  Perhaps the server could wait until receiving packets before starting the child shell?

(Definitely a bikeshed.)

cgull commented:
I finally looked at this code a couple of days ago.  The larger question of whether the Mosh project wants to add this sort of functionality is still there, but I thought I should at least review the code and see how it looks.

@kevinr, @rinne:  I reviewed this pull request because it was conveniently lumped together in one diff.  But the Mosh project prefers a clean Git history:  we want to see changes better broken down into more discrete, incremental pieces of work, commits that reflect the initial development process and wouldn't be relevant once brought into the Mosh repo (bug fixes, implementation changes, etc) should be squashed, and in general the commits should tell a clear, readable, reviewable story of how you got from there to here.  I didn't look quite closely enough, but I *think* having everything squashed down into one commit is a better place to start than @rinne's original commits, which have one large commit containing most of the agent-forwarding work (and the MOSH_ESCAPE_KEY feature too) and then a series of bug fixes following from there.  It's not immediately clear to me how you'd break this into incremental pieces of work (which does say something about how tied the various parts are to each other).

But it mostly looks pretty good to me.  The biggest issue that I see is that it integrates awkwardly with Mosh's (imperfect) event handling, and has a couple of problems with blocking I/O.  Everything I noticed is quite fixable, though.

The network stack here is fairly simple (which I think is appropriate).  It has what's needed for agent forwarding.  But its reliable stream implementation is very simple, it's synchronous and unbuffered.  Those of you hoping that this code could bring port forwarding, X11 forwarding, and other higher-performance things like that...this isn't ready for that.

rinne commented:
@cgull Thanks for going through the stuff.

I merged the forwarding branch with original commit history to current upstream master. The resulting branch is ssh-agent-forwarding-ng in https://github.com/rinne/mosh/

The agent work has two main commits. One being the oob layer and another being the agent forwarding on top of that. Probably with some of the history might benefit from rebasing, but I don't think I have time nor energy for that. If someone can spare the time, please do.

What you said about OOB not being ready for generic stream forwarding, I fully agree. Generic streaming actually sets quite a few requirements to intra-protocol flow control and congestion detection (flow control must be fully implemented for each forwarded stream, not just the protocol connection itself). Also for it to have a decent performance, it needs to have some kind of self adjusting windowing etc also in protocol level. I remember all this being such a pain for ssh development, that I feel pretty strongly that implementing generic stream forwarding to mosh would be much more work than it's worth. And even with someone willing to invest the effort, it would have stability issues and even possible security implications that would take a huge effort to solve. I just don't think it's worth it.

However, it was practically free to implement OOB so that it has separate modes (unreliable datagram mode, reliable datagram mode, reliable stream mode) at least in API level. It should more or less work in either mode, but agent connection is forwarded using reliable datagram mode (one agent protocol packet per datagram) simply in order to avoid inherent problems of generic streaming. It does however mean, that agent protocol packets have to be parsed in mosh-server just enough to read their length and the corresponding payload.

As I said earlier, I feel quite comfortable with current "works for me" -level of mosh agent forwarding implementation. It's not just hacked together up to the point it's somewhat usable, but as far as I can see, it doesn't have any obvious problems and also the way it's hooked into mosh has been at least somewhat designed and not just somehow scotchtaped together. I'm sure that could have been done also in another way, but at least I'm not going to re-implement that. I think it's useful functionality and use it myself every single day, so in one form or another, I hope it makes its way to mosh upstream.

Kriechi commented:
@kevinr if this PR is still considered active, would you mind rebasing and including recent changes from master? Thanks.

rinne commented:
I merged OOB + agent forwarding on top of current master and did another PR #696.

cgull commented:
I'd like to consider this for Mosh 1.3, after we release 1.2.6.  I've started a discussion on the mosh-devel mailing list; if you're interested please comment there.

cgull commented:
I'd like to consider this for Mosh 1.3, after we release 1.2.6.  I've started a discussion on the mosh-devel mailing list; if you're interested please comment there.

OneOfOne commented:
This is very important for us with crappy connections that need to transfer files over ssh.

andersk commented:
@OneOfOne: File transfer is outside the scope of this patch (and probably of Mosh, for the foreseeable future).

kevinr commented:
@cgull I'd love to get it merged in 1.3.  I don't see your thread on the mosh-devel list, though.

andersk commented:
@kevinr http://mailman.mit.edu/pipermail/mosh-devel/2016-May/001297.html

cgull commented:
@kevinr: I got a bounce from Gmail when I sent to mosh-devel, which I think is related to recent DMARC changes they've made, which I believe will cause some mailing list mail (including from my domain) to bounce.  If you're using Gmail that's probably why you didn't see it.

ErwanAliasr1 commented:
Any chance to get that feature in ?

rinne commented:
@ErwanAliasr1 See #696 instead. This one should be closed anyways.

Mic92 commented:
I also have some goodies merged in here. https://github.com/Mic92/mosh/commits/agent-forwarding
I rebase this fork every once in a while

rinne commented:
> > @ErwanAliasr1 See #696 instead. This one should be closed anyways.
> 
> To be clear to passers-by like me who might otherwise get excited: @andersk explained #696 [only covers only SSH agent forwarding](https://github.com/mobile-shell/mosh/pull/696#issuecomment-420900696) (and has been merged 🎉), not all out-of-band data (port forwarding).
> 
> It's unfortunate that half of this was dropped, it's probably a lot of work now to fit the pieces into the latest code. Is it possible that's what's happening on [this fork](https://github.com/rinne/mosh) by @rinne?

Hi.  Actually none of that has been merged to upstream. #696 is just another thread of the same thing because of something I don't remember any longer. Anyways, there OOB data is not port forwarding as such, but it's just a protocol level extension that could be used in port forwarding like it's now used with agent forwarding. Anyways, none of these have made itself into upstream. I merge the upstream master to agent enabled branch (see #696) usually when I have to compile Mosh for myself, which happens maybe every few months or so.

rinne commented:
And OOB stuff as it is now isn't really suitable for generic port forwarding. It lacks flow control and should there be a connection in which the reader side is blocking and producer side keeps writing, it simply closes the connection.
